prometheus/echoprometheus/prometheus.go

```diff
@@ -0,0 +1,458 @@
+					status = http.StatusInternalServerError
+				}
+			}
+
+			values := make([]string, len(labelNames))
+			values[0] = strconv.Itoa(status)
+			values[1] = c.Request().Method
+			values[2] = c.Request().Host
+			values[3] = url
+			for _, cv := range customValuers {
+				values[cv.index] = cv.valueFunc(c, err)
+			}
+
+			requestDuration.WithLabelValues(values...).Observe(elapsed)
+			requestCount.WithLabelValues(values...).Inc()
+			requestSize.WithLabelValues(values...).Observe(float64(reqSz))
+			responseSize.WithLabelValues(values...).Observe(float64(c.Response().Size))
+
+			return err
+		}
+	}, nil
+}
+
+type customLabelValuer struct {
+	index     int
+	label     string
+	valueFunc LabelValueFunc
+}
+
+func createLabels(customLabelFuncs map[string]LabelValueFunc) ([]string, []customLabelValuer) {
+	labelNames := []string{"code", "method", "host", "url"}
+	if len(customLabelFuncs) == 0 {
+		return labelNames, nil
+	}
+
+	customValuers := make([]customLabelValuer, 0)
+	// we create valuers in two passes for a reason - first to get fixed order, and then we know to assign correct indexes
+	for label, labelFunc := range customLabelFuncs {
+		customValuers = append(customValuers, customLabelValuer{
+			label:     label,
+			valueFunc: labelFunc,
+		})
+	}
+	sort.Slice(customValuers, func(i, j int) bool {
+		return customValuers[i].label < customValuers[j].label
+	})
+
+	for cvIdx, cv := range customValuers {
+		idx := containsAt(labelNames, cv.label)
+		if idx == -1 {
+			idx = len(labelNames)
+			labelNames = append(labelNames, cv.label)
+		}
+		customValuers[cvIdx].index = idx
+	}
+	return labelNames, customValuers
+}
+
+func containsAt[K comparable](haystack []K, needle K) int {
+	for i, v := range haystack {
+		if v == needle {
+			return i
+		}
+	}
+	return -1
+}
+
+func computeApproximateRequestSize(r *http.Request) int {
+	s := 0
+	if r.URL != nil {
+		s = len(r.URL.Path)
+	}
+
+	s += len(r.Method)
+	s += len(r.Proto)
+	for name, values := range r.Header {
+		s += len(name)
+		for _, value := range values {
+			s += len(value)
+		}
+	}
+	s += len(r.Host)
+
+	// N.B. r.Form and r.MultipartForm are assumed to be included in r.URL.
+
+	if r.ContentLength != -1 {
+		s += int(r.ContentLength)
+	}
+	return s
+}
+
+// RunPushGatewayGatherer starts pushing collected metrics and waits for it context to complete or ErrorHandler to return error.
+//
+// Example:
+// ```
+//
+//	go func() {
+//		config := echoprometheus.PushGatewayConfig{
+//			PushGatewayURL: "https://host:9080",
+//			PushInterval:   10 * time.Millisecond,
+//		}
+//		if err := echoprometheus.RunPushGatewayGatherer(context.Background(), config); !errors.Is(err, context.Canceled) {
+//			log.Fatal(err)
+//		}
+//	}()
+//
+// ```
+func RunPushGatewayGatherer(ctx context.Context, config PushGatewayConfig) error {
+	if config.PushGatewayURL == "" {
+		return errors.New("push gateway URL is missing")
+	}
+	if config.PushInterval <= 0 {
+		config.PushInterval = 1 * time.Minute
```

Comment:
Same comment here about mergo would apply if you choose to use mergo in this package